File: frontend/src/lib/planner.ts
import type { Task, PlanItem, NextAction, RadarItem } from './types'

const dayMs = 86400000
export const todayISO = () => { const d = new Date(); return d.toISOString().slice(0, 10) }
export function daysUntil(due?: string | null): number | null {
  if (!due) return null
  const t = new Date(todayISO() + 'T12:00:00').getTime()
  const d = new Date(due.slice(0, 10) + 'T12:00:00').getTime()
  return (d - t) / dayMs + 0.5
}
export const remaining = (t: Task) => Math.max(0, Math.round(t.estimated_minutes * (1 - t.progress / 100)))

function urgency(days: number | null): number {
  if (days === null) return 8
  if (days < 0) return 40
  if (days <= 1) return 38
  if (days <= 2) return 32
  if (days <= 3) return 27
  if (days <= 5) return 22
  if (days <= 7) return 18
  if (days <= 14) return 10
  return 5
}

export function priorityScore(t: Task): { score: number; why: string[] } {
  if (t.status === 'done' || t.progress >= 100) return { score: 0, why: ['Completed.'] }
  const rem = remaining(t), days = daysUntil(t.due_date)
  const u = urgency(days)
  const effort = Math.min(15, (rem / 180) * 15)
  const diff = t.difficulty * 3, imp = t.importance * 3
  let pressure = 0
  if (days !== null && days >= 0) pressure = Math.min(10, (rem / Math.max(1, (days + 1) * 60)) * 10)
  const score = Math.max(1, Math.min(100, u + effort + diff + imp + pressure))
  const why: string[] = []
  if (days === null) why.push('No due date set, so urgency is low')
  else if (days < 0) why.push('Overdue — needs immediate attention')
  else if (days <= 1) why.push('Due within 24 hours')
  else if (days <= 3) why.push(`Due in ~${Math.ceil(days)} days`)
  else why.push(`Due in ~${Math.ceil(days)} days`)
  if (rem >= 150) why.push(`Large workload (~${rem} min left) — start early`)
  else if (rem <= 30) why.push(`Quick win (~${rem} min left)`)
  if (t.difficulty >= 4) why.push('High difficulty — needs focused time')
  if (t.importance >= 4) why.push('High importance for your grade')
  if (pressure >= 6) why.push('Workload density is high for the time left')
  return { score: Math.round(score * 10) / 10, why }
}

export function riskOf(t: Task): RadarItem {
  const rem = remaining(t), days = daysUntil(t.due_date)
  if (t.status === 'done' || t.progress >= 100)
    return { task_id: t.id, title: t.title, level: 'Safe', reason: 'Completed.', recommendation: 'Keep the streak going.' }
  if (days === null)
    return { task_id: t.id, title: t.title, level: 'Safe', reason: 'No deadline set.', recommendation: 'Set a due date so Deadline Radar can protect it.' }
  if (days < 0)
    return { task_id: t.id, title: t.title, level: 'At Risk', reason: `Overdue with ~${rem} min still remaining.`, recommendation: 'Do a 25-minute focus session right now.' }
  if (days <= 2.5 && rem >= 120 && t.progress < 50)
    return { task_id: t.id, title: t.title, level: 'At Risk', reason: `${rem} min estimated, ${Math.max(1, Math.round(days))} days left, only ${t.progress}% complete.`, recommendation: 'Start a 45-minute research session today.' }
  if (days <= 3 && rem >= 180)
    return { task_id: t.id, title: t.title, level: 'At Risk', reason: `${rem} min estimated with ~${Math.ceil(days)} days remaining.`, recommendation: 'Split it into smaller sessions starting today.' }
  const cap = 180 * Math.max(0.5, days)
  if (rem / cap > 0.5 || (days <= 5 && t.progress < 30 && rem > 90))
    return { task_id: t.id, title: t.title, level: 'Watch', reason: `${rem} min left over ~${Math.max(1, Math.ceil(days))} days.`, recommendation: 'Schedule one focused session in the next 2 days.' }
  return { task_id: t.id, title: t.title, level: 'Safe', reason: `${rem} min left with ~${Math.max(1, Math.ceil(days))} days to go.`, recommendation: 'On track — keep it in the daily plan.' }
}

export function ranked(tasks: Task[]) {
  return tasks
    .map((t) => ({ task: t, ...priorityScore(t), risk: riskOf(t) }))
    .sort((a, b) => b.score - a.score)
}

export function dailyPlan(tasks: Task[], hours = 3): { plan: PlanItem[]; warnings: string[] } {
  const open = tasks.filter((t) => t.status !== 'done' && t.progress < 100)
  const r = ranked(open)
  const budget = Math.round(hours * 60)
  const plan: PlanItem[] = []
  const warnings: string[] = []
  let used = 0, mins = 16 * 60
  const fmt = (m: number) => `${String(Math.floor(m / 60)).padStart(2, '0')}:${String(m % 60).padStart(2, '0')}`
  for (const x of r) {
    const rem = remaining(x.task)
    if (rem <= 0) continue
    const chunk = Math.min(rem, rem > 45 ? 45 : rem)
    if (used + chunk > budget) {
      warnings.push(`Not enough time today for all of '${x.task.title}' — remainder moved to tomorrow.`)
      continue
    }
    plan.push({ task_id: x.task.id, title: x.task.title, subject: x.task.subject, start: fmt(mins), minutes: chunk, priority_score: x.score, why: x.why.slice(0, 2).join('; ') })
    used += chunk; mins += chunk + 10
  }
  if (used > budget * 0.95) warnings.push("Today is packed — try 'Rescue My Week' to rebalance.")
  return { plan, warnings }
}

export function nextAction(tasks: Task[]): NextAction {
  const open = tasks.filter((t) => t.status !== 'done' && t.progress < 100)
  if (!open.length) return { title: "You're all caught up!", why: 'No open tasks. Review notes or get ahead.', minutes: 20, task_id: null }
  const top = ranked(open)[0]
  const rem = remaining(top.task)
  const sub = top.task.subtasks.find((s) => !s.done)?.title
  const minutes = Math.min(45, Math.max(20, Math.min(rem, rem > 60 ? 25 : rem)))
  return {
    title: sub ? `Spend the next ${minutes} minutes on '${sub}' (${top.task.title}).` : `Spend the next ${minutes} minutes on '${top.task.title}' (${top.task.subject}).`,
    why: top.why.slice(0, 2).join('; '), minutes, task_id: top.task.id,
    priority_score: top.score, risk: top.risk.level
  }
}

export interface Rescue { total_hours: number; busiest_day: string; busiest_minutes: number; overloaded_by: number; moves: string[]; per_day: number }
export function rescueWeek(tasks: Task[], hours = 3): Rescue {
  const open = tasks.filter((t) => t.status !== 'done' && t.progress < 100)
  const total = open.reduce((a, t) => a + remaining(t), 0)
  const byDay: Record<string, number> = {}
  open.forEach((t) => { const k = t.due_date ?? 'unscheduled'; byDay[k] = (byDay[k] ?? 0) + remaining(t) })
  const entries = Object.entries(byDay).sort((a, b) => b[1] - a[1])
  const busiest_day = entries[0]?.[0] ?? '—', busiest_minutes = entries[0]?.[1] ?? 0
  const moves: string[] = []
  for (const t of [...open].sort((a, b) => remaining(a) - remaining(b))) {
    const rem = remaining(t)
    if (rem >= 120) moves.push(`Split '${t.title}' into ${Math.max(2, Math.min(4, Math.floor(rem / 60)))} sessions (~${Math.floor(rem / Math.max(2, Math.min(4, Math.floor(rem / 60))))} min each).`)
    else if (t.importance <= 2) moves.push(`Moved '${t.title}' earlier — low priority, flexible.`)
    if (moves.length >= 4) break
  }
  if (!moves.length && open.length) moves.push(`Rebalanced '${open[0].title}' to protect your lightest day.`)
  moves.push('Friday evening protected.')
  return { total_hours: Math.round((total / 60) * 10) / 10, busiest_day, busiest_minutes, overloaded_by: Math.max(0, busiest_minutes - Math.round(hours * 60)), moves: moves.slice(0, 5), per_day: Math.round(total / 5) }
}

File: backend/app/planner.py
"""Adaptive Academic Planner — the core innovation of StudyFlow AI.

NOT a todo-list sort by deadline. Priority fuses:
  deadline urgency + remaining effort + difficulty + importance
  + workload pressure + progress + dependency readiness.

Deterministic, explainable, and used as the Demo Intelligence fallback
when no AI API key is configured.
"""
from __future__ import annotations
from datetime import datetime, date, timedelta
from typing import Optional
import math
import uuid

from .models import Task


def _parse_due(due: Optional[str], today: date) -> Optional[date]:
    if not due:
        return None
    try:
        return date.fromisoformat(due[:10])
    except ValueError:
        return None


def days_until(task: Task, today: date) -> Optional[float]:
    d = _parse_due(task.due_date, today)
    if d is None:
        return None
    return (d - today).days + 0.5  # assume mid-day granularity


def remaining_minutes(task: Task) -> int:
    rem = task.estimated_minutes * (1 - task.progress / 100)
    return max(0, int(round(rem)))


def deadline_urgency_score(days: Optional[float]) -> float:
    """0-40 scale."""
    if days is None:
        return 8.0
    if days < 0:
        return 40.0
    if days <= 1:
        return 38.0
    if days <= 2:
        return 32.0
    if days <= 3:
        return 27.0
    if days <= 5:
        return 22.0
    if days <= 7:
        return 18.0
    if days <= 14:
        return 10.0
    return 5.0


def priority_score(task: Task, today: date, all_ids: set[str] | None = None) -> tuple[float, list[str]]:
    """Return (score 0-100, reasoning bullets)."""
    if task.status == "done" or task.progress >= 100:
        return 0.0, ["Completed — no priority."]
    rem = remaining_minutes(task)
    days = days_until(task, today)
    urgency = deadline_urgency_score(days)
    effort = min(15.0, rem / 180 * 15)  # 3h+ caps out
    difficulty = task.difficulty * 3.0  # 3-15
    importance = task.importance * 3.0  # 3-15
    # workload pressure: big remaining work with little time left
    pressure = 0.0
    if days is not None and days >= 0:
        density = rem / max(1.0, (days + 1) * 60)
        pressure = min(10.0, density * 10)
    # blocked by unfinished prerequisites?
    blocked_penalty = 0.0
    reasons: list[str] = []
    if task.depends_on and all_ids is not None:
        # planner caller marks done tasks; here just note dependency
        reasons.append(f"Depends on {len(task.depends_on)} prerequisite(s).")
    score = urgency + effort + difficulty + importance + pressure - blocked_penalty
    score = max(1.0, min(100.0, score))

    # human-readable reasoning (short, student-friendly)
    if days is None:
        reasons.append("No due date set, so urgency is low.")
    elif days < 0:
        reasons.append("Overdue — needs immediate attention.")
    elif days <= 1:
        reasons.append("Due within 24 hours.")
    elif days <= 3:
        reasons.append(f"Due in ~{int(math.ceil(days))} days.")
    else:
        reasons.append(f"Due in ~{int(math.ceil(days))} days.")
    if rem >= 150:
        reasons.append(f"Large workload (~{rem} min left) — start early.")
    elif rem <= 30:
        reasons.append(f"Quick win (~{rem} min left).")
    if task.difficulty >= 4:
        reasons.append("High difficulty — needs focused time.")
    if task.importance >= 4:
        reasons.append("High importance for your grade.")
    if pressure >= 6:
        reasons.append("Workload density is high for the time left.")
    return round(score, 1), reasons


def risk_level(task: Task, today: date, daily_capacity_min: int = 180) -> tuple[str, str, str]:
    """Return (Safe|Watch|At Risk, reason, recommendation)."""
    if task.status == "done" or task.progress >= 100:
        return "Safe", "Completed.", "Nice work — keep the streak going."
    rem = remaining_minutes(task)
    days = days_until(task, today)
    if days is None:
        return "Safe", "No deadline set.", "Set a due date so Deadline Radar can protect it."
    if days < 0:
        return "At Risk", f"Overdue with ~{rem} min still remaining.", "Do a 25-minute focus session right now."
    capacity = max(30, daily_capacity_min) * max(0.5, days)
    ratio = rem / capacity if capacity else 1
    # hard rule: big task, little time, low progress
    if days <= 2.5 and rem >= 120 and task.progress < 50:
        return ("At Risk",
                f"{rem} min estimated, {days:.0f} days left, only {task.progress}% complete.",
                "Start a 45-minute session today to break the back of it.")
    if days <= 3 and rem >= 180:
        return ("At Risk",
                f"{rem} min estimated with ~{days:.0f} days remaining.",
                "Split it into smaller sessions starting today.")
    if ratio > 0.5 or (days <= 5 and task.progress < 30 and rem > 90):
        return ("Watch",
                f"{rem} min left over ~{max(1, int(math.ceil(days)))} days.",
                "Schedule one focused session in the next 2 days.")
    return ("Safe",
            f"{rem} min left with ~{max(1, int(math.ceil(days)))} days to go.",
            "On track — keep it in the daily plan.")


def prioritize(tasks: list[Task], today: date) -> list[dict]:
    ids = {t.id for t in tasks}
    ranked = []
    for t in tasks:
        score, why = priority_score(t, today, ids)
        level, reason, rec = risk_level(t, today)
        ranked.append({"task": t, "score": score, "why": why,
                       "risk": level, "risk_reason": reason, "recommendation": rec})
    ranked.sort(key=lambda r: r["score"], reverse=True)
    return ranked


def auto_subtasks(title: str, estimated_minutes: int, subject: str) -> list[dict]:
    """Break large assignments into actionable steps."""
    if estimated_minutes < 90:
        return []
    steps_templates = [
        "Choose topic & scope", "Gather sources", "Create outline",
        "Draft first version", "Complete analysis / core work",
        "Review & revise", "Final polish & submit",
    ]
    n = 4 if estimated_minutes < 150 else 5 if estimated_minutes < 240 else 7
    steps = steps_templates[:n]
    per = max(20, estimated_minutes // n)
    return [{"id": f"st-{uuid.uuid4().hex[:6]}", "title": s,
             "estimated_minutes": per, "done": False} for s in steps]


def generate_daily_plan(tasks: list[Task], today: date, available_hours: float = 3.0) -> tuple[list[dict], list[str]]:
    ranked = prioritize([t for t in tasks if t.status != "done" and t.progress < 100], today)
    budget = int(available_hours * 60)
    plan, warnings = [], []
    used = 0
    start_hour = 16  # after school default
    cursor = datetime(today.year, today.month, today.day, start_hour, 0)
    for r in ranked:
        t: Task = r["task"]
        rem = remaining_minutes(t)
        if rem <= 0:
            continue
        chunk = min(rem, 45 if rem > 45 else rem)  # 45-min max blocks
        if used + chunk > budget:
            warnings.append(f"Not enough time today for all of '{t.title}' — moved remainder to tomorrow.")
            continue
        plan.append({
            "task_id": t.id, "title": t.title, "subject": t.subject,
            "start": cursor.strftime("%H:%M"), "minutes": chunk,
            "priority_score": r["score"], "why": "; ".join(r["why"][:2]),
        })
        used += chunk
        cursor += timedelta(minutes=chunk + 10)  # 10-min break
    if used > budget * 0.95:
        warnings.append("Today is packed — consider a 'Rescue My Week' rebalance.")
    return plan, warnings


def next_action(tasks: list[Task], today: date) -> dict:
    ranked = prioritize([t for t in tasks if t.status != "done" and t.progress < 100], today)
    if not ranked:
        return {"title": "You're all caught up!",
                "why": "No open tasks. Review notes or get ahead.",
                "minutes": 20, "task_id": None}
    top = ranked[0]
    t: Task = top["task"]
    rem = remaining_minutes(t)
    # first undone subtask becomes the concrete action
    subtask_title = None
    for s in (t.subtasks or []):
        d = s.done if hasattr(s, "done") else s.get("done")
        title = s.title if hasattr(s, "title") else s.get("title")
        if not d:
            subtask_title = title
            break
    minutes = min(45, max(20, min(rem, 25 if rem > 60 else rem)))
    action = f"Spend the next {minutes} minutes on '{subtask_title}' ({t.title})." if subtask_title \
        else f"Spend the next {minutes} minutes on '{t.title}' ({t.subject})."
    return {"title": action, "why": "; ".join(top["why"][:2]),
            "minutes": minutes, "task_id": t.id,
            "priority_score": top["score"], "risk": top["risk"]}


def rescue_week(tasks: list[Task], today: date, available_hours: float = 3.0) -> dict:
    """Detect overloaded days, split big tasks, return before/after."""
    open_tasks = [t for t in tasks if t.status != "done" and t.progress < 100]
    total_min = sum(remaining_minutes(t) for t in open_tasks)
    capacity_week = int(available_hours * 60 * 5)
    # naive before: everything crammed near deadlines
    per_day_before: dict[str, int] = {}
    for t in open_tasks:
        key = t.due_date or "unscheduled"
        per_day_before[key] = per_day_before.get(key, 0) + remaining_minutes(t)
    busiest_day = max(per_day_before.items(), key=lambda kv: kv[1]) if per_day_before else ("—", 0)
    overloaded_by = max(0, busiest_day[1] - int(available_hours * 60))
    moves, splits = [], []
    for t in sorted(open_tasks, key=lambda x: remaining_minutes(x)):
        rem = remaining_minutes(t)
        if rem >= 120:
            parts = max(2, min(4, rem // 60))
            splits.append(f"Split '{t.title}' into {parts} sessions (~{rem // parts} min each).")
        elif t.importance <= 2:
            moves.append(f"Moved '{t.title}' earlier — low priority, flexible.")
    if len(moves) + len(splits) == 0 and open_tasks:
        moves.append(f"Rebalanced '{open_tasks[0].title}' to protect your lightest day.")
    after_days = 5
    per_day_after = round(total_min / after_days) if after_days else 0
    return {
        "total_hours": round(total_min / 60, 1),
        "days": 5,
        "before": {"busiest_day": busiest_day[0], "busiest_minutes": busiest_day[1],
                   "label": f"{total_min/60:.1f}h concentrated workload"},
        "after": {"per_day_minutes": per_day_after,
                  "label": "Balanced workload across 5 days"},
        "overloaded_by_minutes": overloaded_by,
        "moves": (moves + splits)[:5],
        "protected": "Friday evening protected.",
    }


def weekly_review(tasks: list[Task], sessions: list[dict] | None = None) -> dict:
    done = [t for t in tasks if t.status == "done" or t.progress >= 100]
    total = len(tasks)
    pct = round(len(done) / total * 100) if total else 0
    study_min = sum(s.get("minutes", 25) for s in (sessions or []))
    insights = [
        f"You completed {pct}% of your planned work this week.",
        "Starting long assignments 1–2 days earlier would smooth your Fridays.",
    ]
    if done:
        insights.insert(1, f"Strongest finish: '{done[0].title}'.")
    return {"completion_rate": pct, "study_minutes": study_min,
            "completed": len(done), "total": total, "insights": insights}
